fast_llm/engine/training/config.py

```diff
@@ -17,48 +25,85 @@
+    )
+
+    def _validate(self):
+        if self.interval:
+            self.offset %= self.interval
+        super()._validate()
+
+    def enabled(self, iteration: int | None = None):
+        return self.interval and (iteration is None or (iteration - self.offset) % self.interval == 0)
+
+    def is_sub_interval(self, other: "IntervalConfig"):
+        if not self.enabled():
+            return True
+        elif not other.enabled():
+            return False
+        return self.interval % other.interval == 0 and (other.offset % other.interval) == (
+            self.offset % other.interval
         )
 
-        def enabled(self, iteration: int | None = None):
-            return self.interval and (iteration is None or (iteration - self.offset) % self.interval == 0)
+    def assert_sub_interval(self, other: "IntervalConfig"):
+        assert self.is_sub_interval(other), f"{self} is not a sub-interval of {other}"
 
-        def is_sub_interval(self, other: "IntervalConfig"):
-            if not self.enabled():
-                return True
-            elif not other.enabled():
-                return False
-            return self.interval % other.interval == 0 and (other.offset % other.interval) == (
-                self.offset % other.interval
-            )
+    def get_count(self, iteration):
+        # Number of times this interval was enabled after a given iteration.
+        return (iteration - self.offset) // self.interval + 1 if self.enabled() else 0
 
-        def assert_sub_interval(self, other: "IntervalConfig"):
-            assert self.is_sub_interval(other), f"{self} is not a sub-interval of {other}"
 
-    return IntervalConfig
+def _validate_script(value):
+    if isinstance(value, str):
+        value = shlex.split(value)
+    Assert.geq(len(value), 1)
+    return value
 
 
 @config_class()
-class WandbAlertConfig(
-    get_interval_config_class(
-        "Wandb status post (alert). Must be a multiple of the logging interval",
-        "Wandb status post (alert). Must be compatible with the logging offset",
+class CallbackConfig(Config):
+    script: list[str] | None = Field(
+        default=None,
+        desc="Shell script to run.",
+        hint=FieldHint.feature,
+        valid=skip_valid_if_none(_validate_script),
+    )
+    environment: dict[str, str] = Field(
+        default_factory=dict,
+        desc="Environment variables to add to the script.",
+        hint=FieldHint.feature,
+    )
+
+    def run(self):
+        if self.script is not None:
+            environment = os.environ.copy()
+            environment.update(self.environment)
+            subprocess.Popen(self.script, env=environment)
+
+
+@config_class()
+class WandbAlertConfig(IntervalConfig):
+    interval = FieldUpdate(
+        desc="The number of training iterations between each Wandb status post (alert)."
+        " Setting to None will disable iteration-based wandb alerts."
+        " Must be a sub-interval of the logging interval."
```

Comment:
you mean that wandb posting can only happen at logging times? That would be super-interval, no? can you clarify what you mean here? I'm confused.

Reply:
Yes. By sub-interval I meant posting iterations are a subset of logging ones, but that's probably not the right term.